### PR TITLE
Mimir query engine: emit annotations for binary operations between incompatible native histograms

### PR DESCRIPTION
**Pull request by charleskorn**
#### What this PR does

This PR adds the same annotations as https://github.com/grafana/mimir/pull/8924, this time for binary operations such as addition and subtraction.

#### Which issue(s) this PR fixes or relates to

(none)

#### Checklist

- [x] Tests updated.
- [n/a] Documentation added.
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`.
- [n/a] [`about-versioning.md`](https://github.com/grafana/mimir/blob/main/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

### CODE REVIEW

File: pkg/streamingpromql/engine_test.go [hunk 1]
```diff
@@ -1444,6 +1444,21 @@ func TestAnnotations(t *testing.T) {
 			},
 		},
 
+		"binary operation between native histograms with exponential and custom buckets": {
+			data: nativeHistogramsWithCustomBucketsData,
+			expr: `metric{series="exponential-buckets"} + ignoring(series) metric{series="custom-buckets-1"}`,
+			expectedWarningAnnotations: []string{
+				`PromQL warning: vector contains a mix of histograms with exponential and custom buckets schemas for metric name "" (1:1)`,
```
Comment (charleskorn) [inline pkg/streamingpromql/engine_test.go:1451]: Note to reviewers: these messages look weird given we have no metric name in the case of binary operations, but it's what Prometheus does as well, so I've kept it the same for the sake of consistency.

Comment (jhesketh) [reply pkg/streamingpromql/engine_test.go:1451]: We should open an upstream issue to fix it there and then eventually in MQE

File: pkg/streamingpromql/operators/binary_operation.go [hunk 2]
```diff
@@ -812,7 +822,13 @@ func (b *BinaryOperation) computeResult(left types.InstantVectorSeriesData, righ
 			// Timestamps match at this step
 			resultFloat, resultHist, ok, err := b.opFunc(lF, rF, lH, rH)
 			if err != nil {
-				return types.InstantVectorSeriesData{}, err
+				err = functions.NativeHistogramErrorToAnnotation(err, b.emitAnnotation)
+				if err == nil {
+					// Error was converted to an annotation, continue without emitting a sample here.
+					ok = false
```
Comment (jhesketh) [inline pkg/streamingpromql/operators/binary_operation.go:828]: (nit) I think `ok` will always be false when there is an error. No harm in being explicit though.

**Review by jhesketh — APPROVED** [review]

File: pkg/streamingpromql/operators/binary_operation.go (hunk 2 above)
Comment (charleskorn) [reply pkg/streamingpromql/operators/binary_operation.go:828]: Yep, that's exactly my thinking.

File: pkg/streamingpromql/engine_test.go (hunk 1 above)
Comment (charleskorn) [reply pkg/streamingpromql/engine_test.go:1451]: Yep, done: https://github.com/prometheus/prometheus/issues/14669

**Pull request merged**